File: internal/repository/additionalStatusRepository.go
package repository

import (
	"database/sql"
	"log"

	"Student-management-system/internal/models"
)

type AdditionalStatusRepository struct {
	db *sql.DB
}

func NewAdditionalStatusRepository(db *sql.DB) *AdditionalStatusRepository {
	return &AdditionalStatusRepository{db: db}
}

func (r *AdditionalStatusRepository) GetByID(id int) (*models.AdditionalStatus, error) {
	query := "SELECT id, name FROM additional_statuses WHERE id = $1"
	row := r.db.QueryRow(query, id)

	var status models.AdditionalStatus
	err := row.Scan(&status.ID, &status.Name)
	if err != nil {
		log.Printf("Ошибка при получении статуса: %v", err)
		return nil, err
	}

	return &status, nil
}

func (r *AdditionalStatusRepository) Create(status *models.AdditionalStatus) error {
	query := "INSERT INTO additional_statuses (name) VALUES ($1) RETURNING id"
	err := r.db.QueryRow(query, status.Name).Scan(&status.ID)
	if err != nil {
		log.Printf("Ошибка создания статуса: %v", err)
		return err
	}

	return nil
}

func (r *AdditionalStatusRepository) Delete(id int) error {
	query := "DELETE FROM additional_statuses WHERE id = $1"
	_, err := r.db.Exec(query, id)
	return err
}

File: app/repositories/additional_status_repository.py
from typing import Optional, List
from asyncpg import Connection
from .base import BaseRepository
from ..models.additional_status import AdditionalStatus, AdditionalStatusCreate, AdditionalStatusUpdate


class AdditionalStatusRepository(BaseRepository[AdditionalStatus]):
    """Репозиторий для работы с дополнительными статусами"""
    
    @property
    def table_name(self) -> str:
        return "additionalstatuses"
    
    @property
    def model_class(self):
        return AdditionalStatus
    
    async def create(self, data: AdditionalStatusCreate, conn: Optional[Connection] = None) -> AdditionalStatus:
        """Создать дополнительный статус"""
        query = """
            INSERT INTO additionalstatuses (name) 
            VALUES ($1) 
            RETURNING *
        """
        
        async with self._get_connection(conn) as connection:
            row = await connection.fetchrow(query, data.name)
            return AdditionalStatus(**dict(row))
    
    async def update(self, id: int, data: AdditionalStatusUpdate, conn: Optional[Connection] = None) -> Optional[AdditionalStatus]:
        """Обновить дополнительный статус"""
        update_data = data.model_dump(exclude_unset=True)
        if not update_data:
            return await self.get_by_id(id, conn)
        
        query = """
            UPDATE additionalstatuses 
            SET name = $2
            WHERE id = $1
            RETURNING *
        """
        
        async with self._get_connection(conn) as connection:
            row = await connection.fetchrow(query, id, data.name)
            return AdditionalStatus(**dict(row)) if row else None
    
    async def get_by_name(self, name: str, conn: Optional[Connection] = None) -> Optional[AdditionalStatus]:
        """Получить статус по имени"""
        query = "SELECT * FROM additionalstatuses WHERE name = $1"
        
        async with self._get_connection(conn) as connection:
            row = await connection.fetchrow(query, name)
            return AdditionalStatus(**dict(row)) if row else None
    
    async def get_by_student_id(self, student_id: int, conn: Optional[Connection] = None) -> List[AdditionalStatus]:
        """Получить статусы студента"""
        query = """
            SELECT a.* FROM additionalstatuses a
            JOIN studentadditionalstatuses sas ON sas.statusid = a.id
            WHERE sas.studentid = $1
        """
        
        async with self._get_connection(conn) as connection:
            rows = await connection.fetch(query, student_id)
            return [AdditionalStatus(**dict(row)) for row in rows]